# Spam Mail Detection using Machine Learning

This project classifies SMS messages as Spam or Ham using NLP and Machine Learning techniques.




In [2]:
import pandas as pd
from sklearn import metrics
import numpy as np
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt


In [3]:
#The dataset contained non-UTF encoded characters, so latin-1 encoding was used while loading.
spam=pd.read_csv('spam.csv',encoding='latin-1')
spam.head(10)

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
0,ham,"Go until jurong point, crazy.. Available only ...",NaN,NaN,NaN
1,ham,Ok lar... Joking wif u oni...,NaN,NaN,NaN
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,NaN,NaN,NaN
3,ham,U dun say so early hor... U c already then say...,NaN,NaN,NaN
4,ham,"Nah I don't think he goes to usf, he lives aro...",NaN,NaN,NaN
5,spam,FreeMsg Hey there darling it's been 3 week's n...,NaN,NaN,NaN
6,ham,Even my brother is not like to speak with me. ...,NaN,NaN,NaN
7,ham,As per your request 'Melle Melle (Oru Minnamin...,NaN,NaN,NaN
8,spam,WINNER!! As a valued network customer you have...,NaN,NaN,NaN
9,spam,Had your mobile 11 months or more? U R entitle...,NaN,NaN,NaN


In [4]:
#Keeping only relevant columns and removing unnecessary ones.
spam=spam[['v1','v2']]
spam.head()

,v1,v2
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


In [5]:
spam.columns=['label','mail']
spam.head(10)

,label,mail
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."
5,spam,FreeMsg Hey there darling it's been 3 week's n...
6,ham,Even my brother is not like to speak with me. ...
7,ham,As per your request 'Melle Melle (Oru Minnamin...
8,spam,WINNER!! As a valued network customer you have...
9,spam,Had your mobile 11 months or more? U R entitle...


In [6]:
#Convert labels ---> numerical values
#ham -> 0 and spam -> 1
spam['label']=spam['label'].map({'ham':0,'spam':1})
spam.head()

,label,mail
0,0,"Go until jurong point, crazy.. Available only ..."
1,0,Ok lar... Joking wif u oni...
2,1,Free entry in 2 a wkly comp to win FA Cup fina...
3,0,U dun say so early hor... U c already then say...
4,0,"Nah I don't think he goes to usf, he lives aro..."


In [7]:
spam.describe().transpose()

,count,mean,std,min,25%,50%,75%,max
label,5572.0,0.134063,0.340751,0.0,0.0,0.0,0.0,1.0


In [8]:
#target column
spam.groupby(['label']).count()

,mail
label,
0,4825
1,747


In [9]:
#split the data into train and test
X=spam['mail']
y=spam['label']
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=42)
print("Training samples:", X_train.shape[0])
print("Testing samples:", X_test.shape[0])

Training samples: 3900
Testing samples: 1672


In [10]:
#Text Vectorization using TF-IDF¶
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer(
    stop_words='english',
    max_features=3000
)

X_train_tfidf = tfidf.fit_transform(X_train)
X_test_tfidf = tfidf.transform(X_test)

print("TF-IDF Train shape:", X_train_tfidf.shape)
print("TF-IDF Test shape:", X_test_tfidf.shape)


TF-IDF Train shape: (3900, 3000)
TF-IDF Test shape: (1672, 3000)


## Model Training and Evaluation

Training and comparing Logistic Regression, Decision Tree, Random Forest, and KNN models.

In [11]:
model = LogisticRegression(max_iter=1000)
#training model
model.fit(X_train_tfidf, y_train)


LogisticRegression(max_iter=1000)

In [12]:
#Predict labels for the test dataset
y_pred = model.predict(X_test_tfidf)

In [24]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report,precision_score,recall_score,f1_score
# Accuracy, Precision, Recall, F1-score and Confusion Matrix

print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall:", recall_score(y_test, y_pred))
print("F1 Score:", f1_score(y_test, y_pred))
print("\nConfusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))


Accuracy: 0.9605263157894737
Precision: 0.9811320754716981
Recall: 0.7123287671232876
F1 Score: 0.8253968253968254

Confusion Matrix:
 [[1450    3]
 [  63  156]]

Classification Report:
               precision    recall  f1-score   support

           0       0.96      1.00      0.98      1453
           1       0.98      0.71      0.83       219

    accuracy                           0.96      1672
   macro avg       0.97      0.86      0.90      1672
weighted avg       0.96      0.96      0.96      1672



In [25]:
from sklearn.tree import DecisionTreeClassifier
dtc=DecisionTreeClassifier()
dtc.fit(X_train_tfidf,y_train)
dtc_pred=dtc.predict(X_test_tfidf)
print('Accuracy = ',accuracy_score(y_test,dtc_pred))
print("Precision:", precision_score(y_test, dtc_pred))
print("Recall:", recall_score(y_test, dtc_pred))
print("F1 Score:", f1_score(y_test, dtc_pred))
print('Confusion Matrix = \n',confusion_matrix(y_test,dtc_pred))
print('Classification Report = \n',classification_report(y_test,dtc_pred))

Accuracy =  0.9677033492822966
Precision: 0.8837209302325582
Recall: 0.867579908675799
F1 Score: 0.8755760368663594
Confusion Matrix = 
 [[1428   25]
 [  29  190]]
Classification Report = 
               precision    recall  f1-score   support

           0       0.98      0.98      0.98      1453
           1       0.88      0.87      0.88       219

    accuracy                           0.97      1672
   macro avg       0.93      0.93      0.93      1672
weighted avg       0.97      0.97      0.97      1672



In [26]:
from sklearn.ensemble import RandomForestClassifier
rf=RandomForestClassifier(n_estimators = 500)
rf.fit(X_train_tfidf,y_train)
rf_pred=rf.predict(X_test_tfidf)
print('Accuracy = ',accuracy_score(y_test,rf_pred))
print("Precision:", precision_score(y_test, rf_pred))
print("Recall:", recall_score(y_test, rf_pred))
print("F1 Score:", f1_score(y_test, rf_pred))
print('Confusion Matrix = \n',confusion_matrix(y_test,rf_pred))
print('Classification Report = \n',classification_report(y_test,rf_pred))

Accuracy =  0.979066985645933
Precision: 0.9893617021276596
Recall: 0.8493150684931506
F1 Score: 0.914004914004914
Confusion Matrix = 
 [[1451    2]
 [  33  186]]
Classification Report = 
               precision    recall  f1-score   support

           0       0.98      1.00      0.99      1453
           1       0.99      0.85      0.91       219

    accuracy                           0.98      1672
   macro avg       0.98      0.92      0.95      1672
weighted avg       0.98      0.98      0.98      1672



In [30]:
from sklearn.neighbors import KNeighborsClassifier
knn=KNeighborsClassifier(n_neighbors=3)
knn.fit(X_train_tfidf,y_train)
knn_pred=knn.predict(X_test_tfidf)
print('Accuracy = ',accuracy_score(y_test,knn_pred))
print("Precision:", precision_score(y_test, knn_pred))
print("Recall:", recall_score(y_test, knn_pred))
print("F1 Score:", f1_score(y_test, knn_pred))
print('Confusion Matrix = \n',confusion_matrix(y_test,knn_pred))
print('Classification Report = \n',classification_report(y_test,knn_pred))

Accuracy =  0.93122009569378
Precision: 1.0
Recall: 0.4748858447488584
F1 Score: 0.6439628482972136
Confusion Matrix = 
 [[1453    0]
 [ 115  104]]
Classification Report = 
               precision    recall  f1-score   support

           0       0.93      1.00      0.96      1453
           1       1.00      0.47      0.64       219

    accuracy                           0.93      1672
   macro avg       0.96      0.74      0.80      1672
weighted avg       0.94      0.93      0.92      1672



## Conclusion

Among all the tested models, Random Forest achieved the highest accuracy of 97.91% and delivered the best overall performance for spam message classification.